File: pkg/go-server/pubsub_local.go
// This file contains the LocalPubSub implementation which provides an in-memory
// publish-subscribe system for single-node deployments. It implements the PubSub
// interface using Go channels and is suitable for development and single-server setups.
package pondsocket

import (
	"context"
	"sync"
)

type LocalPubSub struct {
	mu         sync.RWMutex
	subs       map[string][]subscription
	closed     bool
	ctx        context.Context
	cancel     context.CancelFunc
	bufferSize int
}

type subscription struct {
	pattern string
	handler func(topic string, data []byte)

	ch     chan PubSubMessage
	cancel context.CancelFunc
}

// NewLocalPubSub creates a new local in-memory PubSub implementation.
// The bufferSize parameter sets the channel buffer size for each subscription.
// If bufferSize is <= 0, it defaults to 100.
// The context is used for lifecycle management of the PubSub system.
func NewLocalPubSub(ctx context.Context, bufferSize int) *LocalPubSub {
	if bufferSize <= 0 {
		bufferSize = 100
	}
	pubsubCtx, cancel := context.WithCancel(ctx)

	return &LocalPubSub{
		subs:       make(map[string][]subscription),
		ctx:        pubsubCtx,
		cancel:     cancel,
		bufferSize: bufferSize,
	}
}

// Subscribe registers a handler for messages matching the given pattern.
// Multiple handlers can be registered for the same pattern.
// Each subscription runs in its own goroutine to prevent blocking.
// Returns an error if the PubSub system is closed.
func (l *LocalPubSub) Subscribe(pattern string, handler func(topic string, data []byte)) error {
	l.mu.Lock()

	defer l.mu.Unlock()

	if l.closed {
		return &pubsubClosedError{}
	}
	subCtx, cancel := context.WithCancel(l.ctx)

	ch := make(chan PubSubMessage, l.bufferSize)

	sub := subscription{
		pattern: pattern,
		handler: handler,
		ch:      ch,
		cancel:  cancel,
	}
	l.subs[pattern] = append(l.subs[pattern], sub)

	go l.runSubscription(subCtx, sub)

	return nil
}

func (l *LocalPubSub) runSubscription(ctx context.Context, sub subscription) {
	for {
		select {
		case <-ctx.Done():
			return
		case msg := <-sub.ch:
			go sub.handler(msg.Topic, msg.Data)
		}
	}
}

// Unsubscribe removes all handlers for the given pattern.
// It cancels the contexts and closes the channels for all subscriptions.
// Returns an error if the pattern was not subscribed or if the system is closed.
func (l *LocalPubSub) Unsubscribe(pattern string) error {
	l.mu.Lock()

	defer l.mu.Unlock()

	if l.closed {
		return &pubsubClosedError{}
	}
	subs, exists := l.subs[pattern]
	if !exists {
		return notFound("pubsub", "pattern not found")
	}
	for _, sub := range subs {
		sub.cancel()

		close(sub.ch)
	}
	delete(l.subs, pattern)

	return nil
}

// Publish sends a message to all subscribers whose patterns match the topic.
// The message is sent asynchronously to prevent blocking.
// If a subscriber's channel is full, the message is dropped for that subscriber.
// Returns an error if the PubSub system is closed.
func (l *LocalPubSub) Publish(topic string, data []byte) error {
	l.mu.RLock()

	defer l.mu.RUnlock()

	if l.closed {
		return &pubsubClosedError{}
	}
	msg := PubSubMessage{
		Topic: topic,
		Data:  data,
	}
	for pattern, subs := range l.subs {
		if matchTopic(pattern, topic) {
			for _, sub := range subs {
				select {
				case sub.ch <- msg:
				default:
				}
			}
		}
	}
	return nil
}

// Close shuts down the LocalPubSub system.
// It cancels all subscription contexts and closes all channels.
// After closing, no new subscriptions or publishes are allowed.
// This method is idempotent and can be called multiple times safely.
func (l *LocalPubSub) Close() error {
	l.mu.Lock()

	defer l.mu.Unlock()

	if l.closed {
		return nil
	}
	l.closed = true
	l.cancel()

	for _, subs := range l.subs {
		for _, sub := range subs {
			close(sub.ch)
		}
	}
	l.subs = make(map[string][]subscription)

	return nil
}

File: python/pondsocket/src/pondsocket/pubsub_local.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass

from .errors import not_found
from .pubsub import PubSubClosedError, PubSubHandler, match_topic


@dataclass(slots=True)
class _Subscription:
    pattern: str
    handler: PubSubHandler
    queue: asyncio.Queue[tuple[str, bytes]]
    task: asyncio.Task[None]


class LocalPubSub:
    __slots__ = ("_buffer_size", "_closed", "_lock", "_subs")

    def __init__(self, *, buffer_size: int = 100) -> None:
        if buffer_size <= 0:
            buffer_size = 100
        self._buffer_size = buffer_size
        self._closed = False
        self._lock = asyncio.Lock()
        self._subs: dict[str, list[_Subscription]] = {}

    async def subscribe(self, pattern: str, handler: PubSubHandler) -> None:
        async with self._lock:
            if self._closed:
                raise PubSubClosedError()
            queue: asyncio.Queue[tuple[str, bytes]] = asyncio.Queue(self._buffer_size)
            task = asyncio.create_task(self._run_subscription(queue, handler))
            sub = _Subscription(pattern=pattern, handler=handler, queue=queue, task=task)
            self._subs.setdefault(pattern, []).append(sub)

    async def _run_subscription(
        self,
        queue: asyncio.Queue[tuple[str, bytes]],
        handler: PubSubHandler,
    ) -> None:
        while True:
            try:
                topic, data = await queue.get()
            except asyncio.CancelledError:
                return
            try:
                await handler(topic, data)
            except Exception:
                continue

    async def unsubscribe(self, pattern: str) -> None:
        async with self._lock:
            if self._closed:
                raise PubSubClosedError()
            if pattern not in self._subs:
                raise not_found("pubsub", "pattern not found")
            subs = self._subs.pop(pattern)
        for sub in subs:
            sub.task.cancel()

    async def publish(self, topic: str, data: bytes) -> None:
        if self._closed:
            raise PubSubClosedError()
        async with self._lock:
            targets: list[_Subscription] = []
            for pattern, subs in self._subs.items():
                if match_topic(pattern, topic):
                    targets.extend(subs)
        for sub in targets:
            try:
                sub.queue.put_nowait((topic, data))
            except asyncio.QueueFull:
                continue

    async def close(self) -> None:
        async with self._lock:
            if self._closed:
                return
            self._closed = True
            all_subs = [s for subs in self._subs.values() for s in subs]
            self._subs.clear()
        for sub in all_subs:
            sub.task.cancel()
